### DISCUSSION TITLE
Data filtering attributes

### DISCUSSION THREAD

**Discussion by piotrzajac** [AutoFixture/AutoFixture#1427] [Show and tell]
In the [AutoFixture.XUnit2.AutoMock](https://github.com/ObjectivityLtd/AutoFixture.XUnit2.AutoMock) project we have introduced a set of data filtering attributes. I wonder if it would be worth including them in AutoFixture packages - if so please let me know.

## What is [AutoFixture.XUnit2.AutoMock](https://github.com/ObjectivityLtd/AutoFixture.XUnit2.AutoMock)?
Yet another library of attributs connecting XUnit2, AutoFixture and one of the selected mocking library: Moq, NSubstitute or FakeItEasy.

## Data filtering attributes
The following attributes helps narrowing down data generation to specific values or omitting certain values.

For these attributes to work, they must be used in conjunction with other data generation attributes.

They can be applied to simple types and collections.

#### Except

An attribute ensuring that values from outside the specified list will be generated.

```csharp
[Theory]
[AutoData]
public void ExceptAttributeUsage(
    [Except(DayOfWeek.Saturday, DayOfWeek.Sunday)] DayOfWeek workday)
{
    Assert.True(workday is >= DayOfWeek.Monday and <= DayOfWeek.Friday);
}
```

#### PickFromRange

An attribute ensuring that only values from specified range will be generated.

```csharp
[Theory]
[AutoData]
public void RangeAttributeUsage(
    [PickFromRange(11, 19)] int teenagerAge)
{
    Assert.True(teenagerAge is > 11 and < 19);
}
```

#### PickFromValues

An attribute ensuring that only values from the specified list will be generated.

```csharp
[Theory]
[AutoData]
public void ValuesAttributeUsage(
    [PickFromValues(DayOfWeek.Saturday, DayOfWeek.Sunday)] HashSet<DayOfWeek> weekend)
{
    var weekendDays = new[] { DayOfWeek.Saturday, DayOfWeek.Sunday };
    Assert.Equivalent(weekendDays, weekend);
}
```

**Comment by aivascu** [comment]
@piotrzajac thank you, for the contribution to the AutoFixture community!
The amount of features in the repository is ... impressive. 😅
I'll need some time to review the code, and then I'll open a request issue on your repo, with the things that might be included in the base AutoFixture repository.

**Reply by piotrzajac** [reply]
Thank you and take your time. In case you are considering using other functionalities, e.g. dependabot, github action, analyzers, etc. please let me know. I can help with limited time 😉
